internal/encoding/gotypes: support go import aliasing

phoban01 opened the pull request:
This change enables support for exporting aliased package names from source cue to go code.

Fixes #3974

phoban01 committed:
internal/encoding/gotypes: support go import aliasing

This change enables support for exporting aliased package names from
source cue to go code.

The special case of the "_" import alias is explicitly ignored.

Fixes #3974

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

mvdan commented on cmd/cue/cmd/testdata/script/cmd_exp_gengo_alias.txtar:
```diff
@@ -0,0 +1,46 @@
+
+#ObjectMeta: {}
+
+-- cue.mod/gen/k8s.io/api/core/v1/types_go_gen.cue --
```
could you rewrite this test so it doesn't use `cue.mod/gen`? e.g. make these sub-packages inside the same module.

mvdan commented on internal/encoding/gotypes/generate.go:
```diff
@@ -112,7 +121,12 @@ func Generate(ctx *cue.Context, insts ...*build.Instance) error {
+				if ok {
+					quotedPath = fmt.Sprintf("%s %s", alias, quotedPath)
+				}
+				printf("\t%s\n", quotedPath)
```
I'd avoid the two-layered formatting, like:

```
if alias, ok := g.importAliases[path]; ok {
    printf("\t%s %q\n", alias, path)
} else {
    printf("\t%q\n", path)
}
```

mvdan commented on internal/encoding/gotypes/generate.go:
```diff
@@ -323,6 +340,7 @@ func (g *generator) emitType(val cue.Value, optionalStg optionalStrategy) (typeF
 		if err != nil {
 			return facts, fmt.Errorf("cannot parse @go type expression: %w", err)
 		}
+
```
please avoid whitespace changes elsewhere, as they are more likely to cause git conflicts down the line

mvdan commented on internal/encoding/gotypes/generate.go:
```diff
@@ -679,11 +697,15 @@ func goValueAttr(val cue.Value) cue.Attribute {
 // By default this is the CUE package name, but it can be overriden by a @go() package attribute.
-func goPkgNameForInstance(inst *build.Instance, instVal cue.Value) string {
+// Import aliases will also checked at this point and inserted if found.
+func goPkgNameForInstance(inst *build.Instance, instVal cue.Value, aliasLookup map[string]string) string {
```
name the parameter `importAliases` for consistency.

mvdan commented on internal/encoding/gotypes/generate.go:
```diff
@@ -679,11 +697,15 @@ func goValueAttr(val cue.Value) cue.Attribute {
 // goPkgNameForInstance determines what to name a Go package generated from a CUE instance.
 // By default this is the CUE package name, but it can be overriden by a @go() package attribute.
-func goPkgNameForInstance(inst *build.Instance, instVal cue.Value) string {
+// Import aliases will also checked at this point and inserted if found.
```
I'd reword this slightly, like:
> When supplying importAliases, and if no package attribute is found,
> the returned package name reflects the alias name that the package is being imported as.

mvdan commented on internal/encoding/gotypes/generate.go:
```diff
@@ -100,7 +109,7 @@ func Generate(ctx *cue.Context, insts ...*build.Instance) error {
 		}
 		printf("// Code generated by \"cue exp gengotypes\"; DO NOT EDIT.\n\n")
-		goPkgName := goPkgNameForInstance(inst, instVal)
+		goPkgName := goPkgNameForInstance(inst, instVal, g.importAliases)
```
this is rather crucial - you must NOT use `importAliases` here, so you should use a `nil` map instead. this tracks which Go packages (by name) we generate per directory, as we can only generate one package per directory. This is entirely unrelated to what aliases a package is imported under.

mvdan commented on internal/encoding/gotypes/generate.go:
```diff
@@ -217,6 +231,9 @@ type generator struct {
 	def *generatedDef
+
+	// importAliases maps package names to the alias
+	importAliases map[string]string
```
Note that we organize fields on a per-package and per-definition basis. This should be on a per-package basis I assume.

phoban01 replied:
Not sure I follow here: should I clarify the description or am I missing something in respect of the package handling implementation?

mvdan replied:
I just mean that the fields in this Go struct are grouped by lifetime. `// Fields for each package instance.` groups the fields that are wiped for each new package instance, and `// Fields for each definition.` get wiped for each new definition. So this new field should move to the "per instance" section of the struct fields.

mvdan commented on internal/encoding/gotypes/generate.go:
```diff
@@ -772,3 +794,32 @@ func emitDocs(printf func(string, ...any), name string, groups []*ast.CommentGro
 }
+
+// gatherImportAliases collects the aliases from imports across the instance.
+// We explicitly ignore "_" aliases because CUE does not support.
```
CUE source code should never be using underscore-aliased imports, so I would delete this line as well as the check for `_` below - it just shouldn't happen

mvdan commented on internal/encoding/gotypes/generate.go:
```diff
@@ -772,3 +794,32 @@ func emitDocs(printf func(string, ...any), name string, groups []*ast.CommentGro
+			if !ok {
+				continue
+			}
+			for _, s := range impDecl.Specs {
```
Loop directly over `f.Imports`, which will save you a handful of lines

mvdan commented on internal/encoding/gotypes/generate.go:
```diff
@@ -217,6 +231,9 @@ type generator struct {
 	// def tracks the generation state for a single CUE definition.
 	def *generatedDef
+
+	// importAliases maps package names to the alias
```
this needs clarifying - is it to the _first_ alias that we found for each package? I reckon it should be, but then we should document it as such, and update the code to ignore any second or third alias for the same package path.

but also, what happens if file `f1.cue` does `import same "foo.com/bar/p1"`, and `f2.cue` does `import same "foo.com/bar/p2"`? As far as I can see, you'll generate Go code which imports both packages aliased as `same`.

I think collecting the original aliases is a good idea, because more often than not it will lead to better results - the user will most likely have chosen reasonable aliases. But you'll need some sort of strategy to deal with conflicts, and add a test for that edge case.

I would personally suggest to track which aliases are already taken, and when you encounter an aliased import where the alias is already taken, you can add a number at the end - e.g. the second `same` above could become `same2`.

mvdan replied:
It's also worth noting that you can have duplicates between non-aliased package imports and aliased imports too, so that must also have a test and be checked for duplicates. For example, `import "foo.com/bar/same"` in f1.cue, and `import same "foo.com/bar/p2"` in f2.cue.

mvdan reviewed (changes requested):
Thank you for the patch! This is roughly in line with what I was thinking, and thank you for writing tests. I think a couple of edge cases still need to be ironed out, though.

phoban01 committed:
fix: rewrite test to avoid cue.mod/gen

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 committed:
fix: avoid two-layer format

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 committed:
fix: undo whitespace addition

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 committed:
fix: reword doc string

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 committed:
fix: rename aliasLookup -> importAliases

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 committed:
fix: simplify gatherImportAliases; removing _ alias check

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 committed:
fix: do not pass importAliases when generating goPkgName

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 committed:
fix: ensure conflicting aliases are handled

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 committed:
fix: handle duplicate alias/package names

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 commented:
> Thank you for the patch! This is roughly in line with what I was thinking, and thank you for writing tests. I think a couple of edge cases still need to be ironed out, though.

Thank you for the detailed review @mvdan. I've addressed all the comments, with just the one clarification needed.

mvdan commented on cmd/cue/cmd/testdata/script/cmd_exp_gengo_alias_conflict.txtar:
```diff
@@ -0,0 +1,57 @@
+# reproduction for https://github.com/cue-lang/cue/issues/3974
```
all three testscript files have the same top-level comment. I'd rewrite the comments to summarize which edge case each test covers, like in this case, something like:

    # Two CUE files in a package import two different packages under the same explicit name.

mvdan reviewed (approved):
Thanks! Just one last nit.

phoban01 committed:
test: update header comments

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 committed:
tidy: ensure struct field grouping is respected

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

phoban01 commented:
Thanks @mvdan.

mvdan reviewed (approved):
Thanks!
